File: plugins/advisor-profiles/src/when.ts
import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";

export interface AdvisorWhen {
  files?: string[];
  paths?: string[];
  message_matches?: string;
}

export type ParseWhenResult =
  | { ok: true; when?: AdvisorWhen }
  | { ok: false; error: string };

export type WhenMatchResult =
  | { matches: true }
  | { matches: false; reason: string };

export interface WhenEvaluationContext {
  cwd: string;
  lastUserMessage?: string;
  currentTurnPaths?: string[];
  loadedWatchdogDirs?: string[];
  fileExists?: (filePath: string) => boolean;
}

const ALLOWED_WHEN_KEYS: Record<string, true> = {
  files: true,
  paths: true,
  message_matches: true,
};

export function parseAdvisorWhen(raw: unknown): ParseWhenResult {
  if (raw === undefined || raw === null) {
    return { ok: true, when: undefined };
  }
  if (typeof raw !== "object" || Array.isArray(raw)) {
    return { ok: false, error: "when must be a mapping" };
  }
  const record = raw as Record<string, unknown>;
  for (const key of Object.keys(record)) {
    if (!ALLOWED_WHEN_KEYS[key]) {
      return { ok: false, error: `unknown key in when: "${key}"` };
    }
  }

  let files: string[] | undefined;
  if (record.files !== undefined) {
    if (typeof record.files === "string") {
      files = [record.files.trim()];
    } else if (Array.isArray(record.files)) {
      if (!record.files.every((item) => typeof item === "string")) {
        return { ok: false, error: "when.files must be strings" };
      }
      files = (record.files as string[]).map((s) => s.trim()).filter(Boolean);
    } else {
      return { ok: false, error: "when.files must be a string or list of strings" };
    }
  }

  let paths: string[] | undefined;
  if (record.paths !== undefined) {
    if (typeof record.paths === "string") {
      paths = [record.paths.trim()];
    } else if (Array.isArray(record.paths)) {
      if (!record.paths.every((item) => typeof item === "string")) {
        return { ok: false, error: "when.paths must be strings" };
      }
      paths = (record.paths as string[]).map((s) => s.trim()).filter(Boolean);
    } else {
      return { ok: false, error: "when.paths must be a string or list of strings" };
    }
  }

  let message_matches: string | undefined;
  if (record.message_matches !== undefined) {
    if (typeof record.message_matches !== "string") {
      return { ok: false, error: "when.message_matches must be a string" };
    }
    message_matches = record.message_matches;
  }

  const when: AdvisorWhen = {};
  if (files !== undefined) when.files = files;
  if (paths !== undefined) when.paths = paths;
  if (message_matches !== undefined) when.message_matches = message_matches;
  return { ok: true, when };
}

export function normalizePosixPath(raw: string): string {
  let p = raw.replace(/\\/g, "/").trim();
  if (p.startsWith("./")) {
    p = p.slice(2);
  }
  p = p.replace(/\/+/g, "/");
  if (p.length > 1 && p.endsWith("/")) {
    p = p.slice(0, -1);
  }
  return p;
}

export function globToRegex(pattern: string): RegExp {
  const normalized = pattern.replace(/\\/g, "/");
  let regex = "^";
  let i = 0;
  while (i < normalized.length) {
    if (normalized.startsWith("**/", i)) {
      regex += "(?:.+/)?";
      i += 3;
    } else if (normalized.startsWith("/**", i) && i + 3 === normalized.length) {
      regex += "(?:/.*)?";
      i += 3;
    } else if (normalized.startsWith("**", i)) {
      regex += ".*";
      i += 2;
    } else if (normalized[i] === "*") {
      regex += "[^/]*";
      i += 1;
    } else if (normalized[i] === "?") {
      regex += "[^/]";
      i += 1;
    } else if ("./+^$()[]{}|\\".includes(normalized[i])) {
      regex += "\\" + normalized[i];
      i += 1;
    } else {
      regex += normalized[i];
      i += 1;
    }
  }
  regex += "$";
  return new RegExp(regex);
}

const EXTENSION_RE = /^[A-Za-z0-9_-]+\.[A-Za-z0-9_.-]*[A-Za-z][A-Za-z0-9_.-]*$/;
const DOTFILE_RE = /^\.[A-Za-z0-9_-]+$/;
const NUMBERS_RE = /^[0-9]+(\.[0-9]+)*$/;

export function extractPathTokens(text: string): string[] {
  if (!text) return [];
  const tokens = new Set<string>();
  const matches = text.match(/[A-Za-z0-9_.~/\\-]+/g) ?? [];
  for (const match of matches) {
    let raw = match.replace(/\\/g, "/");
    raw = raw.replace(/^["'(`<\s]+/, "").replace(/[.,:;!?'")>`\s]+$/, "");
    const candidate = normalizePosixPath(raw);
    if (!candidate || candidate === "." || candidate === ".." || candidate === "/") continue;
    if (candidate.startsWith("http://") || candidate.startsWith("https://")) continue;
    if (NUMBERS_RE.test(candidate)) continue;
    const hasSlash = candidate.includes("/");
    const hasExtension = EXTENSION_RE.test(candidate);
    const isDotfile = DOTFILE_RE.test(candidate);
    if (hasSlash || hasExtension || isDotfile) {
      tokens.add(candidate);
    }
  }
  return [...tokens].sort();
}

export function getAncestorDirs(startDir: string): string[] {
  const ancestors: string[] = [];
  let current = path.resolve(startDir);
  const home = path.resolve(os.homedir());
  while (true) {
    ancestors.push(current);
    let hasGit = false;
    try {
      hasGit = fs.existsSync(path.join(current, ".git"));
    } catch {
    }
    if (hasGit || current === home) break;
    const parent = path.dirname(current);
    if (parent === current) break;
    current = parent;
  }
  return ancestors;
}

export function matchAdvisorWhen(
  when: AdvisorWhen | undefined,
  context: WhenEvaluationContext,
): WhenMatchResult {
  if (!when) return { matches: true };
  const hasFiles = when.files !== undefined;
  const hasPaths = when.paths !== undefined;
  const hasMessage = when.message_matches !== undefined;
  if (!hasFiles && !hasPaths && !hasMessage) {
    return { matches: true };
  }

  const fileExists =
    context.fileExists ??
    ((p: string) => {
      try {
        return fs.existsSync(p);
      } catch {
        return false;
      }
    });

  if (hasFiles) {
    const files = when.files ?? [];
    if (files.length === 0) {
      return { matches: false, reason: "when.files: no files configured" };
    }
    const candidateDirs = new Set<string>();
    candidateDirs.add(path.resolve(context.cwd));
    for (const dir of getAncestorDirs(context.cwd)) {
      candidateDirs.add(path.resolve(dir));
    }
    if (context.loadedWatchdogDirs) {
      for (const dir of context.loadedWatchdogDirs) {
        candidateDirs.add(path.resolve(dir));
      }
    }

    let matched = false;
    for (const file of files) {
      if (path.isAbsolute(file)) {
        if (fileExists(file)) {
          matched = true;
          break;
        }
      } else {
        for (const dir of candidateDirs) {
          const fullPath = path.resolve(dir, file);
          if (fileExists(fullPath)) {
            matched = true;
            break;
          }
        }
        if (matched) break;
      }
    }
    if (!matched) {
      return { matches: false, reason: `when.files: none of [${files.join(", ")}] exist` };
    }
  }

  if (hasPaths) {
    const paths = when.paths ?? [];
    if (paths.length === 0) {
      return { matches: false, reason: "when.paths: no paths configured" };
    }
    const turnPaths = (context.currentTurnPaths ?? []).map(normalizePosixPath);
    let matched = false;
    for (const pattern of paths) {
      const regex = globToRegex(pattern);
      if (turnPaths.some((tp) => regex.test(tp))) {
        matched = true;
        break;
      }
    }
    if (!matched) {
      return { matches: false, reason: `when.paths: no paths matched [${paths.join(", ")}]` };
    }
  }

  if (hasMessage) {
    let regex: RegExp;
    try {
      regex = new RegExp(when.message_matches ?? "");
    } catch (error) {
      return {
        matches: false,
        reason: `when.message_matches: invalid regex (${error instanceof Error ? error.message : String(error)})`,
      };
    }
    const userMessage = context.lastUserMessage ?? "";
    if (!regex.test(userMessage)) {
      return {
        matches: false,
        reason: `when.message_matches: user message did not match /${when.message_matches}/`,
      };
    }
  }

  return { matches: true };
}

export function formatWhen(when: AdvisorWhen | undefined): string {
  if (!when) return "always";
  const parts: string[] = [];
  if (when.files && when.files.length > 0) {
    parts.push(`files: ${when.files.join(", ")}`);
  }
  if (when.paths && when.paths.length > 0) {
    parts.push(`paths: ${when.paths.join(", ")}`);
  }
  if (when.message_matches !== undefined) {
    parts.push(`message_matches: /${when.message_matches}/`);
  }
  return parts.length > 0 ? parts.join("; ") : "always";
}

File: plugins/advisor-profiles/hermes/when.py
from __future__ import annotations

from dataclasses import dataclass
import os
from pathlib import Path
import re
from typing import Any, Callable, Dict, List, Optional, Sequence, Set, Tuple


@dataclass(frozen=True)
class AdvisorWhen:
    files: Optional[Tuple[str, ...]] = None
    paths: Optional[Tuple[str, ...]] = None
    message_matches: Optional[str] = None


ALLOWED_WHEN_KEYS = {"files", "paths", "message_matches"}


def parse_advisor_when(raw: Any) -> Tuple[Optional[AdvisorWhen], Optional[str]]:
    if raw is None:
        return None, None
    if not isinstance(raw, dict):
        return None, "when must be a mapping"

    for key in raw.keys():
        if key not in ALLOWED_WHEN_KEYS:
            return None, f'unknown key in when: "{key}"'

    files: Optional[Tuple[str, ...]] = None
    if "files" in raw:
        val = raw["files"]
        if isinstance(val, str):
            files = (val.strip(),)
        elif isinstance(val, list):
            if not all(isinstance(item, str) for item in val):
                return None, "when.files must be strings"
            files = tuple(item.strip() for item in val if item.strip())
        else:
            return None, "when.files must be a string or list of strings"

    paths: Optional[Tuple[str, ...]] = None
    if "paths" in raw:
        val = raw["paths"]
        if isinstance(val, str):
            paths = (val.strip(),)
        elif isinstance(val, list):
            if not all(isinstance(item, str) for item in val):
                return None, "when.paths must be strings"
            paths = tuple(item.strip() for item in val if item.strip())
        else:
            return None, "when.paths must be a string or list of strings"

    message_matches: Optional[str] = None
    if "message_matches" in raw:
        val = raw["message_matches"]
        if not isinstance(val, str):
            return None, "when.message_matches must be a string"
        message_matches = val

    return AdvisorWhen(files=files, paths=paths, message_matches=message_matches), None


def normalize_posix_path(raw: str) -> str:
    p = raw.replace("\\", "/").strip()
    if p.startswith("./"):
        p = p[2:]
    p = re.sub(r"/+", "/", p)
    if len(p) > 1 and p.endswith("/"):
        p = p[:-1]
    return p


def glob_to_regex(pattern: str) -> re.Pattern:
    normalized = pattern.replace("\\", "/")
    regex = "^"
    i = 0
    while i < len(normalized):
        if normalized.startswith("**/", i):
            regex += r"(?:.+/)?"
            i += 3
        elif normalized.startswith("/**", i) and i + 3 == len(normalized):
            regex += r"(?:/.*)?"
            i += 3
        elif normalized.startswith("**", i):
            regex += r".*"
            i += 2
        elif normalized[i] == "*":
            regex += r"[^/]*"
            i += 1
        elif normalized[i] == "?":
            regex += r"[^/]"
            i += 1
        elif normalized[i] in "./+^$()[]{}|\\":
            regex += "\\" + normalized[i]
            i += 1
        else:
            regex += normalized[i]
            i += 1
    regex += "$"
    return re.compile(regex)


_PATH_CANDIDATE_RE = re.compile(r"[A-Za-z0-9_.~/\\-]+")
_EXTENSION_RE = re.compile(r"^[A-Za-z0-9_-]+\.[A-Za-z0-9_.-]*[A-Za-z][A-Za-z0-9_.-]*$")
_DOTFILE_RE = re.compile(r"^\.[A-Za-z0-9_-]+$")
_NUMBERS_RE = re.compile(r"^[0-9]+(\.[0-9]+)*$")


def extract_path_tokens(text: str) -> List[str]:
    if not text:
        return []
    tokens: Set[str] = set()
    for match in _PATH_CANDIDATE_RE.finditer(text):
        raw = match.group(0).replace("\\", "/")
        raw = raw.strip("\"'(`<\r\n\t ").rstrip(".,:;!?'\")>`\r\n\t ")
        candidate = normalize_posix_path(raw)
        if not candidate or candidate in (".", "..", "/"):
            continue
        if candidate.startswith("http://") or candidate.startswith("https://"):
            continue
        if _NUMBERS_RE.match(candidate):
            continue
        has_slash = "/" in candidate
        has_extension = bool(_EXTENSION_RE.match(candidate))
        is_dotfile = bool(_DOTFILE_RE.match(candidate))
        if has_slash or has_extension or is_dotfile:
            tokens.add(candidate)
    return sorted(tokens)


def get_ancestor_dirs(start_dir: Path) -> List[Path]:
    ancestors: List[Path] = []
    current = Path(start_dir).resolve()
    home = Path.home().resolve()
    while True:
        ancestors.append(current)
        has_git = (current / ".git").exists()
        if has_git or current == home:
            break
        if current.parent == current:
            break
        current = current.parent
    return ancestors


def match_advisor_when(
    when: Optional[AdvisorWhen],
    cwd: Path,
    last_user_message: Optional[str] = None,
    current_turn_paths: Optional[Sequence[str]] = None,
    current_turn_text: Optional[str] = None,
    loaded_watchdog_dirs: Optional[Sequence[Path]] = None,
    file_exists: Optional[Callable[[Path], bool]] = None,
) -> Tuple[bool, Optional[str]]:
    if when is None:
        return True, None
    has_files = when.files is not None
    has_paths = when.paths is not None
    has_message = when.message_matches is not None
    if not has_files and not has_paths and not has_message:
        return True, None

    exists = file_exists or (lambda p: Path(p).exists())

    if has_files:
        files = when.files or ()
        if not files:
            return False, "when.files: no files configured"
        candidate_dirs: List[Path] = [Path(cwd).resolve()]
        candidate_dirs.extend(get_ancestor_dirs(Path(cwd)))
        if loaded_watchdog_dirs:
            candidate_dirs.extend(Path(d).resolve() for d in loaded_watchdog_dirs)

        matched = False
        for f in files:
            file_path = Path(f)
            if file_path.is_absolute():
                if exists(file_path):
                    matched = True
                    break
            else:
                for d in candidate_dirs:
                    if exists(d / file_path):
                        matched = True
                        break
                if matched:
                    break
        if not matched:
            return False, f"when.files: none of [{', '.join(files)}] exist"

    if has_paths:
        paths = when.paths or ()
        if not paths:
            return False, "when.paths: no paths configured"
        turn_paths: Set[str] = set()
        if current_turn_paths:
            for p in current_turn_paths:
                turn_paths.add(normalize_posix_path(p))
        if current_turn_text:
            for p in extract_path_tokens(current_turn_text):
                turn_paths.add(normalize_posix_path(p))

        matched = False
        for pattern in paths:
            regex = glob_to_regex(pattern)
            if any(regex.match(tp) for tp in turn_paths):
                matched = True
                break
        if not matched:
            return False, f"when.paths: no paths matched [{', '.join(paths)}]"

    if has_message:
        try:
            regex = re.compile(when.message_matches or "")
        except Exception as exc:
            return False, f"when.message_matches: invalid regex ({exc})"
        msg = last_user_message or ""
        if not regex.search(msg):
            return False, f"when.message_matches: user message did not match /{when.message_matches}/"

    return True, None


def format_when(when: Optional[AdvisorWhen]) -> str:
    if not when:
        return "always"
    parts: List[str] = []
    if when.files:
        parts.append(f"files: {', '.join(when.files)}")
    if when.paths:
        parts.append(f"paths: {', '.join(when.paths)}")
    if when.message_matches:
        parts.append(f"message_matches: /{when.message_matches}/")
    return "; ".join(parts) if parts else "always"
